langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Sort.java

```diff
@@ -18,73 +18,185 @@
 import static org.ballerinalang.util.BLangCompilerConstants.ARRAY_VERSION;
 
 /**
- * Native implementation of lang.array:sort((any|error)[], function).
+ * Native implementation of lang.array:sort((any|error)[], direction, function).
  *
  * @since 1.0
  */
 @BallerinaFunction(
         orgName = "ballerina", packageName = "lang.array", version = ARRAY_VERSION, functionName = "sort",
-        args = {@Argument(name = "arr", type = TypeKind.ARRAY), @Argument(name = "func", type = TypeKind.FUNCTION)},
+        args = {@Argument(name = "arr", type = TypeKind.ARRAY), @Argument(name = "direction", type = TypeKind.OBJECT),
+                @Argument(name = "func", type = TypeKind.OBJECT)},
         returnType = {@ReturnType(type = TypeKind.ARRAY)},
         isPublic = true
 )
 public class Sort {
 
-    public static ArrayValue sort(Strand strand, ArrayValue arr, FPValue<Object, Long> func) {
+    public static ArrayValue sort(Strand strand, ArrayValue arr, Object direction, Object func) {
         checkIsArrayOnlyOperation(arr.getType(), "sort()");
-        ArrayValue aux = new ArrayValueImpl((BArrayType) arr.getType());
-        mergesort(arr, aux, 0, arr.size() - 1, strand, func);
+        FPValue<Object, Object> function = (FPValue<Object, Object>) func;
+        BType elemType = ((BArrayType) arr.getType()).getElementType();
+        boolean isAscending = true;
+        if (direction.toString().equals("descending")) {
+            isAscending = false;
+        }
+
+        Object[][] sortArr = new Object[arr.size()][2];
+        Object[][] sortArrClone = new Object[arr.size()][2];
+        for (int i = 0; i < arr.size(); i++) {
+            if (function != null) {
+                Object x = function.call(new Object[]{strand, arr.get(i), true});
+                sortArr[i][0] = x;
+            } else {
+                sortArr[i][0] = arr.get(i);
+            }
+            sortArr[i][1] = arr.get(i);
+        }
+
+        if (function != null) {
+            elemType = ((BFunctionType) function.getType()).retType;
+        }
+        if (elemType.getTag() == TypeTags.UNION_TAG) {
+            elemType = getType(elemType);
+        }
+        if (elemType.getTag() == TypeTags.ARRAY_TAG) {
+            BType type = ((BArrayType) elemType).getElementType();
+            if (type.getTag() == TypeTags.UNION_TAG) {
+                elemType = getType(type);
+            }
+        }
+
+        mergesort(sortArr, sortArrClone, 0, sortArr.length - 1, isAscending, elemType);
+
+        for (int k = 0; k < sortArr.length; k++) {
+            arr.add(k, sortArr[k][1]);
+        }
+
         return arr;
     }
 
-    // Adapted from https://algs4.cs.princeton.edu/22mergesort/Merge.java.html
-    private static void mergesort(ArrayValue input, ArrayValue aux, int lo, int hi, Strand strand,
-                                  FPValue<Object, Long> comparator) {
+    private static BType getType(BType type) {
```

Comment:
Can we rename the function to indicate what we're actually looking for and change the param to a `BUnionType`?